pkg/reconciler/route/route.go

```diff
@@ -480,10 +447,92 @@ func (c *Reconciler) ensureFinalizer(route *v1alpha1.Route) error {
+	serviceNames, err := c.getServiceNames(ctx, route)
+	if err != nil {
+		return err
+	}
+
+	mainRouteServiceName, err := domains.HostnameFromTemplate(ctx, route.Name, "")
+	if err != nil {
+		return err
+	}
+
+	mainRouteMeta := route.ObjectMeta.DeepCopy()
+	labels.SetVisibility(mainRouteMeta, serviceNames.clusterLocal().Has(mainRouteServiceName))
+
+	host, err := domains.DomainNameFromTemplate(ctx, *mainRouteMeta, route.Name)
+	if err != nil {
+		return err
+	}
+
+	route.Status.URL = &apis.URL{
+		Scheme: "http",
+		Host:   host,
+	}
+
+	return nil
+}
+
+func (c *Reconciler) getServiceNames(ctx context.Context, route *v1alpha1.Route) (serviceNames, error) {
+	// Populate existing service name sets
+	existingServices, err := c.getServices(route)
+	if err != nil {
+		return serviceNames{}, err
+	}
+	existingServiceNames := resources.GetNames(existingServices)
+	existingClusterLocalServices := resources.FilterService(existingServices, resources.IsClusterLocalService)
+	existingClusterLocalServiceNames := resources.GetNames(existingClusterLocalServices)
+	existingPublicServiceNames := existingServiceNames.Difference(existingClusterLocalServiceNames)
+
+	// Populate desired service name sets
+	desiredServiceNames, err := resources.GetDesiredServiceNames(ctx, route)
+	if err != nil {
+		return serviceNames{}, err
+	}
+	desiredPublicServiceNames := desiredServiceNames.Intersection(existingPublicServiceNames)
+	desiredClusterLocalServiceNames := desiredServiceNames.Intersection(existingClusterLocalServiceNames)
+
+	// Any new desired services will follow the default route visibility. We only need to consider the case where it is
+	// "cluster-local". The alternative visibility means it should not be cluster local.
+	serviceWithDefaultVisibility := desiredServiceNames.Difference(existingServiceNames)
+	if labels.IsObjectLocalVisibility(route.ObjectMeta) {
+		desiredClusterLocalServiceNames = desiredClusterLocalServiceNames.Union(serviceWithDefaultVisibility)
+	} else {
+		desiredPublicServiceNames = desiredPublicServiceNames.Union(serviceWithDefaultVisibility)
+	}
+
+	return serviceNames{
+		existingPublicServiceNames:       existingPublicServiceNames,
+		existingClusterLocalServiceNames: existingClusterLocalServiceNames,
+		desiredPublicServiceNames:        desiredPublicServiceNames,
+		desiredClusterLocalServiceNames:  desiredClusterLocalServiceNames,
+	}, nil
+}
+
 /////////////////////////////////////////
 // Misc helpers.
 /////////////////////////////////////////
 
+type serviceNames struct {
+	existingPublicServiceNames       sets.String
+	existingClusterLocalServiceNames sets.String
+	desiredPublicServiceNames        sets.String
+	desiredClusterLocalServiceNames  sets.String
+}
+
+func (sn serviceNames) existing() sets.String {
+	return sn.existingPublicServiceNames.Union(sn.existingClusterLocalServiceNames)
+}
+
+func (sn serviceNames) desired() sets.String {
```

Comment:
It seems desired and existing are used only in this scope where they are unionized, may be create as such from the get go?

Reply:
I'd be happy to remove `desired` for now, as it isn't used at the moment. It feels a bit more flexible to offer it this way rather than storing just `existing` and `clusterlocal`.

After rereading your comment, did you actually mean: calculate the union for those and keep it around in the struct for access?

Reply:
re: https://github.com/knative/serving/pull/4821#discussion_r305428567
Will delete the unused for now